lib/editor/actions/map/stopStrategies.js

```diff
@@ -220,26 +220,42 @@ export function addStopAtInterval (latlng: LatLng, activePattern: Pattern, contr
   }
 }
 
-export function addStopToPattern (pattern: Pattern, stop: GtfsStop, index?: ?number) {
+export function addStopToPattern (pattern: Pattern, stop: GtfsStop | GtfsLocation, index?: ?number) {
+  // eslint-disable-next-line complexity
   return async function (dispatch: dispatchFn, getState: getStateFn) {
     const {data, editSettings} = getState().editor
     const {followStreets} = editSettings.present
-    const {patternStops: currentPatternStops, shapePoints} = pattern
+    const {
+      patternLocations: currentPatternLocations,
+      patternStops: currentPatternStops,
+      shapePoints
+    } = pattern
+    const patternLocations = clone(currentPatternLocations)
     const patternStops = clone(currentPatternStops)
     const {controlPoints, patternSegments} = getControlPoints(getState())
     const patternLine = lineString(coordinatesFromShapePoints(shapePoints))
     const hasShapePoints = shapePoints && shapePoints.length > 2
+    let patternHalts = mergePatternHalts(patternStops, patternLocations)
     const newStop = stopToPatternStop(
       stop,
       (typeof index === 'undefined' || index === null)
-        ? patternStops.length
+        ? patternHalts.length
         : index
     )
 
-    if (typeof index === 'undefined' || index === null || index === patternStops.length) {
-      // Push pattern stop to cloned list.
-      patternStops.push(newStop)
-      if (hasShapePoints) {
+    // $FlowFixMe Flow doesn't understand type check
+    if (newStop.hasOwnProperty('stopId')) patternStops.push(newStop)
+    // $FlowFixMe Flow doesn't understand type check
+    if (newStop.hasOwnProperty('locationId')) patternLocations.push(newStop)
+
+    if (typeof index === 'undefined' || index === null || index === patternHalts.length) {
+      // Checking for stop_lat and stop_lon is how we check if we are dealing with
+      // a stop or a location
+      if (hasShapePoints && !!stop.stop_lat && !!stop.stop_lon) {
+        // Push pattern stop to cloned list.
+        // $FlowFixMe Flow doesn't understand type check
+        patternHalts = mergePatternHalts(patternStops, patternLocations)
```

Comment:
I have seen this exact call a dozen times now, and I think it should be refactored.

Reply:
What would your recommendation be? We can't do it at a higher level since `patternStops` and `patternLocations` change